### PR TITLE
Revision of DCE for Vulkan

### PR DESCRIPTION
**Pull request by Qining**
This version of DCE drops more commands and save more transferring data.

Sits on top of #881 and #901 

TODO:
[x] Remove debugging comments and clean the code
[x] Blurbs for the exported type and methods
[x] More tests for the DCE transformation and the simple trie data sturcture
[x] Solve the problem of synchronization primitives
  Solved by keeping all the synchronization primitives alive.
[x] Handle EndRenderPass
  Solved by marking the corresponding EndRenderPass alive if a
  BeginRenderPass is found alive. However, no new backpropation will be
  initiated from the alive EndRenderPass

### CODE REVIEW

File: gapir/cc/interpreter.cpp [hunk 1]
```diff
@@ -446,7 +446,8 @@ Interpreter::Result Interpreter::interpret(uint32_t opcode) {
             DEBUG_OPCODE_26("ADD", opcode);
             return this->add(opcode);
         case InstructionCode::LABEL:
-            DEBUG_OPCODE_26("LABEL", opcode);
+            GAPID_INFO("LABEL" "(%#010x)", opcode & DATA_MASK26)
```
Comment (Qining) [inline gapir/cc/interpreter.cpp]: This will definitely be changed back soon.

File: gapis/api/transform/dce.go [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+	panic(fmt.Errorf("This transform does not accept input atoms"))
+}
+
+func (t *DCE) Flush(ctx context.Context, out Writer) {
+	log.W(ctx, "DCE Flush called on DCE: %p, %v", t, t)
+	t0 := dCECounter.Start()
+	livenessBoard, _ := t.backPropagate(ctx)
+	dCECounter.Stop(t0)
+	flushedCommands := newFullCommandIndexSet()
+	// for ci := uint64(0); t.footprint.GetBehaviourIndex(ctx,
```
Comment (Qining) [inline gapis/api/transform/dce.go]: Such debugging comments will be removed soon

File: gapis/api/transform/dce.go [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+	// }
+	// }
+
+	for bi := uint64(0); bi <= t.endBehaviourIndex; bi++ {
+		bh := t.footprint.Behaviours[bi]
+		fci := bh.BelongTo
+		if livenessBoard[bi] && len(fci) == 1 && !flushedCommands.contains(fci) {
+			flushedCommands.insert(fci)
+
+			// // log the memory size
```
Comment (Qining) [inline gapis/api/transform/dce.go]: Debugging comments

File: gapis/api/transform/dce.go [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+			t.defUse.AddWrites(bh)
+			t.defUse.AddReads(bh)
+			aliveCommands.insert(fci)
+			// log.W(ctx, "BP: Alive: %v, %v", bi, fci)
+		}
+	}
+	if int(commandsCount) > len(t.footprint.Commands) {
+		commandsCount = uint64(len(t.footprint.Commands))
+	}
+	if commandsCount > uint64(0) {
```
Comment (Qining) [inline gapis/api/transform/dce.go]: Need a nicer way to collect the DCE matrix

File: gapis/api/transform/dce.go [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+}
+
+func (t *DCE) clearMachines(ctx context.Context) {
+	for m, _ := range t.defUse.usedMachines {
+		m.Clear()
+	}
+	t.defUse.usedMachines = map[dependencygraph.DefUseMachine]struct{}{}
+}
+
+func (t *DCE) clear(ctx context.Context) {
```
Comment (Qining) [inline gapis/api/transform/dce.go]: A nicer 'clear' logic may change the logic here.

File: gapis/api/transform/dce.go [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+	if ok {
+		if bv, ok := v.(bool); ok {
+			return bv
+		}
+		return false
+	}
+	return false
+}
+
+type DefUseGraph struct {
```
Comment (Qining) [inline gapis/api/transform/dce.go]: This type is here to distribute commands to their corresponding DefUse checking machine according to their API.

File: gapis/api/transform/dead_code_elimination.go [hunk 7]
```diff
@@ -166,7 +166,7 @@ func (t *DeadCodeElimination) propagateLiveness(ctx context.Context) []bool {
 		deadCodeEliminationDrawLiveCounter.AddInt64(int64(numLiveDraws))
 		deadCodeEliminationDataDeadCounter.AddInt64(int64(deadMem))
 		deadCodeEliminationDataLiveCounter.AddInt64(int64(liveMem))
-		log.D(ctx, "DCE: dead: %v%% %v cmds %v MB %v draws, live: %v%% %v cmds %v MB %v draws",
+		log.W(ctx, "DCE: dead: %v%% %v cmds %v MB %v draws, live: %v%% %v cmds %v MB %v draws",
```
Comment (Qining) [inline gapis/api/transform/dead_code_elimination.go]: Will be changed back soon

File: gapis/api/vulkan/custom_replay.go [hunk 8]
```diff
@@ -1668,6 +1668,19 @@ func (a *RecreateImageData) Mutate(ctx context.Context, s *api.State, b *builder
 
 		if !a.Data.IsNullptr() {
 			imageInfo := imageObject.Info
+
```
Comment (Qining) [inline gapis/api/vulkan/custom_replay.go]: Will be removed.

File: gapis/api/vulkan/custom_replay.go [hunk 9]
```diff
@@ -1668,6 +1668,19 @@ func (a *RecreateImageData) Mutate(ctx context.Context, s *api.State, b *builder
...
+			// ///
```
Comment (Qining) [inline gapis/api/vulkan/custom_replay.go]: Will be removed.

File: gapis/api/vulkan/custom_replay.go [hunk 10]
```diff
@@ -1826,6 +1839,17 @@ func (a *RecreateBufferData) Mutate(ctx context.Context, s *api.State, b *builde
 		bufferInfo := bufferObject.Info
 		cb := CommandBuilder{Thread: a.thread}
 
+		///
+		// isCoherentMem, _ := subIsMemoryCoherent(ctx, a, o, s, GetState(s), a.thread, b, bufferObject.Memory)
```
Comment (Qining) [inline gapis/api/vulkan/custom_replay.go]: Will be removed.

File: gapis/api/vulkan/replay.go [hunk 11]
```diff
@@ -488,14 +493,24 @@ func (a API) Replay(
 	}
 
 	// Prepare data for dead-code-elimination
-	dceInfo := deadCodeEliminationInfo{}
-	if !config.DisableDeadCodeElimination {
-		dg, err := dependencygraph.GetDependencyGraph(ctx)
+	// dceInfo := deadCodeEliminationInfo{}
```
Comment (Qining) [inline gapis/api/vulkan/replay.go]: Dead code for debugging, will be cleaned later.

File: gapis/config/config.go [hunk 12]
```diff
@@ -20,7 +20,7 @@ const (
 	DebugReplay                = false
 	DebugReplayBuilder         = false
 	DisableDeadCodeElimination = false
-	DebugDeadCodeElimination   = false
+	DebugDeadCodeElimination   = true
```
Comment (Qining) [inline gapis/config/config.go]: Will be changed back in the final update

File: gapis/resolve/dependencygraph/depgraph.go [hunk 13]
```diff
@@ -0,0 +1,300 @@
...
+	"github.com/google/gapid/gapis/database"
+)
+
+var footprintBuildCounter = benchmark.GlobalCounters.Duration("footprint.build")
+
+// TODO: Remove this ad-hoc command index type and its logic for comparing once
+// the api.CmdID is replaced with []uint64 like type.
+type FullCommandIndex []uint64
+
+func (fci FullCommandIndex) LessThan(fci2 FullCommandIndex) bool {
```
Comment (Qining) [inline gapis/resolve/dependencygraph/depgraph.go]: Probably should be removed as we now have SubCmdIdx. However, here the command index does include the CmdID of the `vkQueueSubmit`, but SubCmdIdx usually does not.

File: gapis/resolve/dependencygraph/depgraph.go [hunk 14]
```diff
@@ -0,0 +1,300 @@
...
+
+func NewArrayBasedTrie() *ArrayBasedTrie {
+	return &ArrayBasedTrie{
+		hasValue:  false,
+		value:     uint64(0),
+		nextLevel: []*ArrayBasedTrie{},
+	}
+}
+
+const ReservedBehaviourIndex = uint64(0xFFFFFFFFFFFFFFFF)
```
Comment (Qining) [inline gapis/resolve/dependencygraph/depgraph.go]: dead

File: gapis/resolve/dependencygraph/depgraph.go [hunk 15]
```diff
@@ -0,0 +1,300 @@
...
+}
+
+type dummyDefUseMachine struct{}
+
+func (m *dummyDefUseMachine) CheckDef(c DefUseAtom) bool { return true }
+func (m *dummyDefUseMachine) Def(DefUseAtom)             {}
+func (m *dummyDefUseMachine) Use(DefUseAtom)             {}
+func (m *dummyDefUseMachine) Clear()                     {}
+
+type ArrayBasedTrie struct {
```
Comment (Qining) [inline gapis/resolve/dependencygraph/depgraph.go]: A simple trie is needed to store FullCommandIndex indexed values

**Comment by Qining** [comment]
@dsrbecky @ben-clayton @AWoloszyn, This CL is ready for review, PTAL.

File: gapis/api/vulkan/externs.go [hunk 16]
```diff
@@ -277,6 +277,46 @@ func (e externs) readMappedCoherentMemory(memory_handle VkDeviceMemory, offset_i
 		mem.Data.Slice(dstStart+r.Base, dstStart+r.Base+r.Size, l).Copy(
 			e.ctx, U8ᵖ(mem.MappedLocation).Slice(srcStart+r.Base, srcStart+r.Base+r.Size, l), e.cmd, e.s, e.b)
 	}
+
+	// if e.cmd.Extras() == nil || e.cmd.Extras().Observations() == nil {
```
Comment (ben-clayton) [inline gapis/api/vulkan/externs.go]: ???

Comment (Qining) [inline gapis/api/vulkan/externs.go]: OOps, sorry, I forgot this piece of code. It was used to debugging the coherent memory observations...I will remove this.

**Comment by ben-clayton** [comment]
We're a little concerned by the amount of hand-written code here, which ideally would come from the API file. Long term we should look to see how we can push this information into the API.

I'll do a proper review tomorrow - looks like I need a good clear schedule for this one. :wink:

**Comment by ben-clayton** [comment]
Please can you rebase?

**Comment by Qining** [comment]
Rebased. Forgot to push

File: gapis/api/vulkan/footprint_builder.go [hunk 17]
```diff
@@ -0,0 +1,3283 @@
...
+	case *RecreateCmdCopyBuffer:
+		srcBufferData, _ := vb.getBufferData(ctx, bh, cmd.SrcBuffer).(memorySpan)
+		dstBufferData, _ := vb.getBufferData(ctx, bh, cmd.DstBuffer).(memorySpan)
+		src := []dependencygraph.DefUseAtom{}
+		dst := []dependencygraph.DefUseAtom{}
+		count := uint64(cmd.RegionCount)
+		regions := cmd.PRegions.Slice(0, count, l)
+		for i := uint64(0); i < count; i++ {
+			region := regions.Index(i, l).Read(ctx, cmd, s, nil)
+			srcOffset := srcBufferDa
```
Comment (ben-clayton) [inline gapis/api/vulkan/footprint_builder.go]: ???

File: gapis/api/vulkan/read_framebuffer.go [hunk 18]
```diff
@@ -116,8 +116,16 @@ func (t *readFramebuffer) Color(id api.CmdID, width, height, bufferIdx uint32, r
 				res(nil, fmt.Errorf("There have been no previous draws"))
 				return
 			}
+			if lastDrawInfo.Framebuffer == nil {
+				res(nil, fmt.Errorf("There have been no framebuffer"))
```
Comment (ben-clayton) [inline gapis/api/vulkan/read_framebuffer.go]: s/have/has

File: gapis/api/vulkan/read_framebuffer.go [hunk 19]
```diff
@@ -116,8 +116,16 @@ func (t *readFramebuffer) Color(id api.CmdID, width, height, bufferIdx uint32, r
...
+				return
+			}
 
-			imageView := lastDrawInfo.Framebuffer.ImageAttachments[bufferIdx]
+			imageView, ok := lastDrawInfo.Framebuffer.ImageAttachments[bufferIdx]
+			if !ok {
+				res(nil, fmt.Errorf("There have been no attchment %v in the framebuffer", bufferIdx))
```
Comment (ben-clayton) [inline gapis/api/vulkan/read_framebuffer.go]: s/have/has

File: gapis/api/vulkan/replay.go [hunk 20]
```diff
@@ -89,6 +89,11 @@ type deadCodeEliminationInfo struct {
 	deadCodeElimination *transform.DeadCodeElimination
 }
 
+type DCEInfo struct {
```
Comment (ben-clayton) [inline gapis/api/vulkan/replay.go]: need to be public?

File: gapis/resolve/dependencygraph/footprint.go [hunk 21]
```diff
@@ -0,0 +1,239 @@
...
+// Modify records a read and a write operation of the given DefUseAtom to the
+// Behaviour
+func (b *Behaviour) Modify(c DefUseAtom) {
+	b.Read(c)
+	b.Write(c)
+}
+
+// DefUseAtom represents a piece of state which can be altered as commands
+// being executed.
+type DefUseAtom interface {
```
Comment (ben-clayton) [inline gapis/resolve/dependencygraph/footprint.go]: `DefUseCmd`

What does the `Def` bit mean?

File: gapis/resolve/dependencygraph/footprint.go [hunk 22]
```diff
@@ -0,0 +1,239 @@
...
+}
+
+// Modify records a read and a write operation of the given DefUseAtom to the
+// Behaviour
+func (b *Behaviour) Modify(c DefUseAtom) {
+	b.Read(c)
+	b.Write(c)
+}
+
+// DefUseAtom represents a piece of state which can be altered as commands
```
Comment (ben-clayton) [inline gapis/resolve/dependencygraph/footprint.go]: All state can be altered as commands are executed, otherwise it isn't really state. What does this really do?

File: gapis/resolve/dependencygraph/footprint.go [hunk 23]
```diff
@@ -0,0 +1,239 @@
...
+// GetBehaviourIndex returns the index of the last Behaviour in the Footprint
+// which belongs to the command or subcomand indexed by the given SubCmdIdx. In
+// case the SubCmdIdx is invalid or the Behaviour index is not found, error
+// will be logged and uint64(0) will be returned.
+func (fpt *Footprint) GetBehaviourIndex(ctx context.Context,
+	fci api.SubCmdIdx) uint64 {
+	switch len(fci) {
+	case 1,
+		// Only API command ID
+		4,
```
Comment (ben-clayton) [inline gapis/resolve/dependencygraph/footprint.go]: I've seen these magic constants repeated a few places. Would it make sense to declare them as named constants?

File: gapis/resolve/dependencygraph/footprint.go [hunk 24]
```diff
@@ -0,0 +1,239 @@
...
+		fpt.Behaviours = append(fpt.Behaviours, b)
+		fpt.BehaviourIndices[b] = bi
+		return true
+	default:
+		log.E(ctx, "Invalid length of SubCmdIdx (full command index): %v", len(fci))
+		return false
+	}
+}
+
+// FootprintBuilderAPI provides FootprintBuilder
```
Comment (ben-clayton) [inline gapis/resolve/dependencygraph/footprint.go]: `FootprintBuilderProvider` ?

File: gapis/resolve/dependencygraph/footprint.go [hunk 25]
```diff
@@ -0,0 +1,239 @@
...
+type FootprintBuilderAPI interface {
+	FootprintBuilder(context.Context) FootprintBuilder
+}
+
+// FootprintBuilder builds Footprint from a stream of commands.
```
Comment (ben-clayton) [inline gapis/resolve/dependencygraph/footprint.go]: `from a stream of commands` I only see one command being passed to the function. Please clarify.

File: gapis/resolve/dependencygraph/footprint.go (hunk 22 above)
Comment (Qining) [inline gapis/resolve/dependencygraph/footprint.go]: `DefUseAtom` is the same thing to the GLES dce's `StateKey` + `StateAddress`. It represents one part of the whole state. Basically it is the basic element on which we track read/write (or say, def/use) operations.

The side effects of the commands are recorded as a list of read/write operations on some `DefUseAtom`, and stored in `Behaviours`

Side effect of executing command X:
* `Behaviour` 0
 * Read `DefUseAtom` A
 * Read `DefUseAtom` B
 * Write `DefUseAtom` C
 * Write `DefUseAtom` D
 * ...
* `Behaviour` 1
 * Read ...
 * Write ...
* ...
* `Behaviour` N

All the `DefUseAtom` together represents the whole state.

File: gapis/api/vulkan/replay.go (hunk 20 above)
Comment (Qining) [inline gapis/api/vulkan/replay.go]: Nope, will change it to private.

File: gapis/resolve/dependencygraph/footprint.go (hunk 21 above)
Comment (Qining) [inline gapis/resolve/dependencygraph/footprint.go]: I was thinking of def-use chain. So as the basic element of a def-use chain, I named it to DefUseAtom. This is actually not related to our api.Cmd, as there is no one-to-one matching. So I guess I should not name it `DefUseCmd` either.

File: gapis/resolve/dependencygraph/footprint.go (hunk 24 above)
Comment (Qining) [inline gapis/resolve/dependencygraph/footprint.go]: (－‸ლ) It was the original name, I will change it back.

File: gapis/api/vulkan/footprint_builder.go [hunk 26]
```diff
@@ -0,0 +1,3283 @@
...
+			return true
+		}
+	}
+	return false
+}
+
+func (m *vulkanMachine) RecordBehaviourEffects(behaviourIndex uint64,
+	ft *dependencygraph.Footprint) []uint64 {
+	bh := ft.Behaviours[behaviourIndex]
+	aliveIndices := []uint64{}
```
Comment (Qining) [inline gapis/api/vulkan/footprint_builder.go]: cool, this looks much better.

File: gapis/api/vulkan/footprint_builder.go (hunk 17 above)
Comment (Qining) [inline gapis/api/vulkan/footprint_builder.go]: I thought this is not supported. I definitely want to group them together.

Comment (Qining) [inline gapis/api/vulkan/footprint_builder.go]: Will change all such cases.

File: gapis/resolve/dependencygraph/footprint.go (hunk 23 above)
Comment (Qining) [inline gapis/resolve/dependencygraph/footprint.go]: I will remove them. It is actually not valid to have Vulkan specific values here.

File: gapis/resolve/dependencygraph/footprint.go [hunk 27]
```diff
@@ -0,0 +1,224 @@
...
+		commandIndexToBehaviourIndexLookupTable: &api.SubCmdIdxTrie{},
+		BehaviourIndices:                        map[*Behaviour]uint64{},
+	}
+}
+
+// Behaviour contains a set of read and write operations as side effect of
+// executing the command to whom it belongs. Behaviour also contains a refernce
+// to the back-propagation machine which should be used to process the
+// Behaviour to determine its liveness for dead code elimination.
+type Behaviour struct {
```
Comment (ben-clayton) [inline gapis/resolve/dependencygraph/footprint.go]: Please use American spelling: `Behavior`

File: gapis/resolve/dependencygraph/footprint.go [hunk 28]
```diff
@@ -0,0 +1,224 @@
...
+	return &Footprint{
+		Commands:                                cmds,
+		Behaviours:                              make([]*Behaviour, 0, len(cmds)),
+		commandIndexToBehaviourIndexLookupTable: &api.SubCmdIdxTrie{},
+		BehaviourIndices:                        map[*Behaviour]uint64{},
+	}
+}
+
+// Behaviour contains a set of read and write operations as side effect of
+// executing the command to whom it belongs. Behaviour also contains a refernce
```
Comment (ben-clayton) [inline gapis/resolve/dependencygraph/footprint.go]: s/refernce/reference

File: gapis/resolve/dependencygraph/footprint.go [hunk 29]
```diff
@@ -0,0 +1,224 @@
...
+// to the back-propagation machine which should be used to process the
+// Behaviour to determine its liveness for dead code elimination.
+type Behaviour struct {
+	Reads    []DefUseAtom
+	Writes   []DefUseAtom
+	BelongTo api.SubCmdIdx
```
Comment (ben-clayton) [inline gapis/resolve/dependencygraph/footprint.go]: Owner?

File: gapis/resolve/dependencygraph/footprint.go (hunk 22 above)
Comment (ben-clayton) [inline gapis/resolve/dependencygraph/footprint.go]: So it refers to a [DU Chain](https://en.wikipedia.org/wiki/Use-define_chain) or something similar?
At the moment we just have an interface with a single blank member and there's no clear description to its intended usage. Things we should be clarifying to the reader:

* What types should implement the interface?
* What uses the interface?
* Why is the only method a single blank function? Is it supposed to be a tag?
* What separates the type from a `interface{}`?

File: gapis/resolve/dependencygraph/footprint.go [hunk 30]
```diff
@@ -0,0 +1,224 @@
...
+)
+
+var footprintBuildCounter = benchmark.GlobalCounters.Duration("footprint.build")
+
+// Footprint contains a list of command and a list of Behaviours which
+// describes the side effect of executing the commands in that list.
+type Footprint struct {
+	Commands                                []api.Cmd
+	Behaviours                              []*Behaviour
+	commandIndexToBehaviourIndexLookupTable *api.SubCmdIdxTrie
```
Comment (ben-clayton) [inline gapis/resolve/dependencygraph/footprint.go]: This isn't java. `cmdIdxToBehavior` will suffice.
Also it probably doesn't need to be a pointer type.
Please put private members after public ones.

File: gapis/resolve/dependencygraph/footprint.go [hunk 31]
```diff
@@ -0,0 +1,224 @@
...
+	return []uint64{behaviourIndex}
+}
+
+func (m *dummyMachine) Clear() {}
+
+// GetBehaviourIndex returns the index of the last Behaviour in the Footprint
+// which belongs to the command or subcomand indexed by the given SubCmdIdx. In
+// case the SubCmdIdx is invalid or a valid Behaviour index is not found, error
+// will be logged and uint64(0) will be returned.
+func (fpt *Footprint) GetBehaviourIndex(ctx context.Context,
```
Comment (ben-clayton) [inline gapis/resolve/dependencygraph/footprint.go]: Unless there's a horrible variable collision, please use single letter method receivers.
Please try to keep function declarations on a single line unless they're crazy long - I find the newline hard to separate from the body.
Also try to avoid `Get` prefixes:

`func (f *Footprint) BehaviorIndex(ctx context.Context, i api.SubCmdIdx) uint64 {`

File: gapis/resolve/dependencygraph/footprint.go [hunk 32]
```diff
@@ -0,0 +1,224 @@
...
+	FootprintBuilder(context.Context) FootprintBuilder
+}
+
+// FootprintBuilder incrementally builds Footprint one command by one command.
+type FootprintBuilder interface {
+	BuildFootprint(context.Context, *api.State, *Footprint, api.CmdID, api.Cmd)
+}
+
+// GetFootprint returns a pointer to the resolved Footprint.
+func GetFootprint(ctx context.Context) (*Footprint, error) {
```
Comment (ben-clayton) [inline gapis/resolve/dependencygraph/footprint.go]: Exception to the `Get` rule here - there's clearly a `Footprint` type declared in this package, so this is fine.

File: gapis/resolve/dependencygraph/footprint.go [hunk 33]
```diff
@@ -0,0 +1,224 @@
...
+	r, err := database.Build(ctx, &FootprintResolvable{Capture: capture.Get(ctx)})
+	if err != nil {
+		return nil, fmt.Errorf("Counld not get execution foot print: %v", err)
+	}
+	return r.(*Footprint), nil
+}
+
+// Resolve implements the database.Resolver interface.
+func (r *FootprintResolvable) Resolve(ctx context.Context) (interface{},
+	error) {
```
Comment (ben-clayton) [inline gapis/resolve/dependencygraph/footprint.go]: nit: rm newline.

File: gapis/resolve/dependencygraph/footprint.go [hunk 34]
```diff
@@ -0,0 +1,224 @@
...
+		return nil, err
+	}
+	cmds := c.Commands
+	builders := map[api.API]FootprintBuilder{}
+
+	ft := NewFootprint(ctx, cmds)
+
+	s := c.NewState()
+	t0 := footprintBuildCounter.Start()
+	for i, cmd := range cmds {
```
Comment (ben-clayton) [inline gapis/resolve/dependencygraph/footprint.go]: Use `api.ForeachCmd`. This is required to prevent `Mutate` from panicing if the context is cancelled.

File: gapis/resolve/dependencygraph/footprint.go [hunk 35]
```diff
@@ -0,0 +1,224 @@
...
+	if err != nil {
+		return nil, err
+	}
+	cmds := c.Commands
+	builders := map[api.API]FootprintBuilder{}
+
+	ft := NewFootprint(ctx, cmds)
+
+	s := c.NewState()
+	t0 := footprintBuildCounter.Start()
```
Comment (ben-clayton) [inline gapis/resolve/dependencygraph/footprint.go]: use: `defer footprintBuildCounter.Stop(t0)`.
You have returns in this function.

File: gapis/resolve/dependencygraph/footprint.go [hunk 36]
```diff
@@ -0,0 +1,224 @@
...
+				// API does not provide execution footprint info, always keep commands
+				// fromt such APIs alive.
+				bh := NewBehaviour(api.SubCmdIdx{uint64(i)}, &dummyMachine{})
+				bh.Alive = true
+				// Even if the command does not belong to an API that provides
+				// execution footprint info, we still need to mutate it in the new
+				// state, because following commands in other APIs may depends on the
+				// side effect of the this command.
+				if err := cmd.Mutate(ctx, s, nil); err != nil {
+					log.W(ctx, "Command %v %v: %v", api.CmdID(i), cmd, err)
```
Comment (ben-clayton) [inline gapis/resolve/dependencygraph/footprint.go]: Looks like this could spam for traces that have issues. The report view should be the place to display these warnings to the user, not the log.

File: gapis/resolve/dependencygraph/footprint.go (hunk 21 above)
Comment (ben-clayton) [inline gapis/resolve/dependencygraph/footprint.go]: Okay, but let's steer clear of `Atom`.
`DefUseElement`, `DefUseNode`, `DefUseData`, `DefUseState`?

File: gapis/resolve/dependencygraph/footprint.go [hunk 37]
```diff
@@ -0,0 +1,224 @@
...
+// Behaviour
+func (b *Behaviour) Modify(c DefUseAtom) {
+	b.Read(c)
+	b.Write(c)
+}
+
+// DefUseAtom represents a piece of state which can be altered as commands
+// being executed.
+type DefUseAtom interface {
+	DefUseAtom()
```
Comment (ben-clayton) [inline gapis/resolve/dependencygraph/footprint.go]: If this is just for tagging types to be used for the dep-graph, then we have to question why we're creating a pointless interface and not using `interface{}`. Does the type implementing this interface actually do anything special, or is it just arbitrary data?

If we can justify the interface, then we've been using the prefix `Is` for these methods: `IsDefUseXXX()`.

File: gapis/api/vulkan/footprint_builder.go [hunk 38]
```diff
@@ -0,0 +1,3283 @@
...
+		handleLoadOp(ctx, bh, r)
+	}
+	for _, i := range qei.subpasses[subpassI].inputAttachments {
+		handleLoadOp(ctx, bh, i)
+	}
+	if qei.subpasses[subpassI].depthStencilAttachment != nil {
+		dsAtt := qei.subpasses[subpassI].depthStencilAttachment
+		// TODO: Not all subpasses change layouts
+		modify(ctx, bh, dsAtt.layout)
+		if (dsAtt.desc.LoadOp == VkAttachmentLoadOp_VK_ATTACHMENT_LOAD_OP_CLEAR ||
```
Comment (Qining) [inline gapis/api/vulkan/footprint_builder.go]: I'm wondering to which file I should add such helpers. Should I add them in the API file as subroutines (but cpp codes is not used), or in the externs.go (make more sense but still far away from the uses), or just in the footprint_builder.go (and make it just a private helper)?

File: gapis/resolve/dependencygraph/footprint.go (hunk 37 above)
Comment (Qining) [inline gapis/resolve/dependencygraph/footprint.go]: This is actually arbitrary data. The only requirement is that the API's backpropagation machine should know how to handle those data. And DefUseAtom() is just a label to tag the data that can be processed by the machine.

I will change it to IsDefUseElement()

File: gapis/resolve/dependencygraph/footprint.go (hunk 21 above)
Comment (Qining) [inline gapis/resolve/dependencygraph/footprint.go]: I changed my mind, it will be renamed to DefUseVariable to be more consistent with the Wiki page: https://en.wikipedia.org/wiki/Live_variable_analysis

File: gapis/api/vulkan/footprint_builder.go (hunk 38 above)
Comment (Qining) [inline gapis/api/vulkan/footprint_builder.go]: I will just make this private functions in this footprint_builder.go for now.

File: gapis/api/vulkan/footprint_builder.go (hunk 17 above)
Comment (Qining) [inline gapis/api/vulkan/footprint_builder.go]: Nope, we cannot do that. This switch is a type switch, we cannot access the members of the concrete api.Cmd. Actually I was thinking of this before, as all the vkCmdxxx  and RecreateCmdxxx commands have exactly the same code. But unfortunately, we cannot do that.

File: gapis/api/vulkan/footprint_builder.go [hunk 39]
```diff
@@ -0,0 +1,3283 @@
...
+			}
+		}
+		swCount := uint64(info.SwapchainCount)
+		vkSws := info.PSwapchains.Slice(0, swCount, l)
+		imgIds := info.PImageIndices.Slice(0, swCount, l)
+		for i := uint64(0); i < swCount; i++ {
+			vkSw := vkSws.Index(i, l).Read(ctx, cmd, s, nil)
+			read(ctx, bh, vkHandle(vkSw))
+			imgId := imgIds.Index(i, l).Read(ctx, cmd, s, nil)
+			{
```
Comment (Qining) [inline gapis/api/vulkan/footprint_builder.go]: removed

File: gapis/api/vulkan/footprint_builder.go [hunk 40]
```diff
@@ -0,0 +1,3283 @@
...
+	if execInfo.subpasses[subpassI].depthStencilAttachment != nil {
+		dsAtt := execInfo.subpasses[subpassI].depthStencilAttachment
+		modify(ctx, bh, dsAtt.data)
+	}
+}
+
+func (vb *VulkanFootprintBuilder) readBoundIndexBuffer(ctx context.Context,
+	bh *dependencygraph.Behaviour, execInfo *queueExecutionInfo, cmd api.Cmd) {
+	boundIndexBufferData, _ := execInfo.currentCmdBufState.indexBuffer.backingData.(memorySpan)
+	indexSize := uint64(2)
```
Comment (Qining) [inline gapis/api/vulkan/footprint_builder.go]: helper function added.

**Comment by Qining** [comment]
All prior comments are resolved.

File: gapis/api/vulkan/footprint_builder.go [hunk 41]
```diff
@@ -424,24 +424,30 @@ func (qei *queueExecutionInfo) updateCurrentCommand(ctx context.Context,
...
 	subpassI := qei.subpass.val
-	handleLoadOp := func(ctx context.Context, bh *dependencygraph.Behaviour,
+	handleLoadOp := func(ctx context.Context, bh *dependencygraph.Behavior,
 		attachment *subpassAttachmentInfo) {
 		// TODO: Not all subpasses change layouts
 		modify(ctx, bh, attachment.layout)
-		switch attachment.desc.LoadOp {
-		case VkAttachmentLoadOp_VK_ATTACHMENT_LOAD_OP_CLEAR,
-			VkAttachmentLoadOp_VK_ATTACHMENT_LOAD_OP_DONT_CARE:
+		if attachment.desc.LoadOp.isLoad() {
```
Comment (ben-clayton) [inline gapis/api/vulkan/footprint_builder.go]: This change doesn't seem equivalent. Intentional?

File: gapis/api/vulkan/footprint_builder.go [hunk 42]
```diff
@@ -457,17 +463,13 @@ func (qei *queueExecutionInfo) startSubpass(ctx context.Context,
 		dsAtt := qei.subpasses[subpassI].depthStencilAttachment
 		// TODO: Not all subpasses change layouts
 		modify(ctx, bh, dsAtt.layout)
-		if (dsAtt.desc.LoadOp == VkAttachmentLoadOp_VK_ATTACHMENT_LOAD_OP_CLEAR ||
-			dsAtt.desc.LoadOp == VkAttachmentLoadOp_VK_ATTACHMENT_LOAD_OP_DONT_CARE) &&
-			(dsAtt.desc.StencilLoadOp == VkAttachmentLoadOp_VK_ATTACHMENT_LOAD_OP_CLEAR ||
-				dsAtt.desc.StencilLoadOp == VkAttachmentLoadOp_VK_ATTACHMENT_LOAD_OP_DONT_CARE) {
+		if dsAtt.desc.LoadOp.isLoad() && dsAtt.desc.StencilLoadOp.isLoad() {
```
Comment (ben-clayton) [inline gapis/api/vulkan/footprint_builder.go]: This change doesn't seem equivalent. Intentional?

File: gapis/api/vulkan/footprint_builder.go (hunk 17 above)
Comment (ben-clayton) [inline gapis/api/vulkan/footprint_builder.go]: I wanted to check the original code, but for some reason I can't... these comments have no source...

Comment (ben-clayton) [inline gapis/api/vulkan/footprint_builder.go]: Found it! Weird github issue.

**Comment by ben-clayton** [comment]
LGTM. Thanks for addressing all the comments.

**Comment by Qining** [comment]
Will merge after all tests on my side pass

**Comment by Qining** [comment]
All tests passed, merged.
